File: python/Problem47.py
# Problem: 47. Permutations II
# URL: https://leetcode.com/problems/permutations-ii/
# Difficulty: Medium
# Tags: Array, Permutation, Backtracking

from typing import List

class Solution:
  def permuteUnique(self, nums: List[int]) -> List[List[int]]:
    count: dict = {n:0 for n in nums}
    results: List[List[int]] = []
    permutations: List[int] = []
    for n in nums:
      count[n] += 1
    def backtrack():
      if len(permutations) == len(nums):
        results.append(permutations[:])
        return
      for n in count:
        if count[n] > 0:
          permutations.append(n)
          count[n] -= 1
          backtrack()
          count[n] += 1
          permutations.pop()
    backtrack()
    return results

File: cpp/Problem47.cpp
/*
 * Problem: 47. Permutations II
 * URL: https://leetcode.com/problems/permutations-ii/
 * Difficulty: Medium
 * Tags: Array, Permutation, Backtracking
 */

#include <algorithm>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
	vector<vector<int>> permuteUnique(vector<int>& nums) {
		for (int n : nums) {
			counts[n]++;
		}
		vector<vector<int>> results;
		backtrack(results, nums);
		return results;
	}

private:
	unordered_map<int, int> counts;
	vector<int> permutations;
	void backtrack(vector<vector<int>> &results, vector<int> &nums) {
		if (permutations.size() == nums.size()) {
			results.push_back(permutations);
			return;
		}
		for (auto itr = counts.begin(); itr != counts.end(); itr++) {
			if (itr->second > 0) {
				permutations.push_back(itr->first);
				itr->second -= 1;
				backtrack(results, nums);
				itr->second += 1;
				permutations.pop_back();
			}
		}
	}
};